### ISSUE TITLE
Performance of Foldable and Traversable instances for List type

### ISSUE THREAD

**Issue by harendra-kumar** [composewell/streamly#132]
The performance of `Foldable` and `Traversable` for `Streamly.List` type does not seem to be the best possible as `id` shows `32 us` and `foldl'` shows `800 us` whereas `sum` and `mapM` show `8 ms` time. We can perhaps do better by defining the instances manually, defining each operation in Foldable/Traversable manually instead of using default implementations.

**Comment by harendra-kumar** [comment]
All of these must have the same performance, more or less:

```
serially/pure/foldl'                     mean 45.43 μs  ( +- 1.133 μs  )
serially/pure/foldable/foldl'            mean 936.6 μs  ( +- 40.27 μs  )
serially/pure/foldable/sum               mean 9.067 ms  ( +- 373.3 μs  )
```

**Comment by pranaysashank** [comment]
@harendra-kumar I believe this issue can now be closed?

**Comment by harendra-kumar** [comment]
This is fixed by #416
